File: src/handlers/transactionHandler.js
const { Markup } = require('telegraf');

class TransactionHandler {
  constructor(db, bot) {
    this.db = db;
    this.bot = bot;
  }

  // Get status emoji
  getStatusEmoji(status) {
    const emojis = {
      'pending': '⏳',
      'paid': '💰',
      'shipped': '📦',
      'delivered': '✅',
      'completed': '🎉',
      'cancelled': '❌',
      'disputed': '⚠️'
    };
    return emojis[status] || '❓';
  }

  // Initiate purchase
  async initiatePurchase(ctx, productId) {
    await ctx.answerCbQuery();

    const product = this.db.getProduct(productId);

    if (!product) {
      return ctx.reply('❌ Product not found or no longer available.');
    }

    if (product.seller_id === ctx.from.id) {
      return ctx.reply('❌ You cannot buy your own product.');
    }

    if (product.status !== 'active') {
      return ctx.reply('❌ This product is no longer available.');
    }

    const seller = this.db.getUser(product.seller_id);
    const sellerName = seller.username ? `@${seller.username}` : seller.first_name;

    this.db.setSessionState(ctx.from.id, 'purchase_form_confirm', {
      productId: product.id
    });

    await ctx.reply(
      '🛒 *Purchase Confirmation*\n\n' +
      `*Product:* ${product.title}\n` +
      `*Description:* ${product.description || 'N/A'}\n` +
      `*Price:* $${product.price.toFixed(2)}\n` +
      `*Seller:* ${sellerName}\n\n` +
      '⚠️ *Important:*\n' +
      '• Funds will be held in escrow\n' +
      '• Release only after delivery confirmation\n' +
      '• Admin mediates any disputes\n\n' +
      'Type "confirm" or "yes" to proceed, or anything else to cancel:',
      { 
        parse_mode: 'Markdown',
        ...Markup.keyboard([['Confirm', 'Cancel']]).resize()
      }
    );
  }

  // Show transaction details
  async showTransactionDetails(ctx, txId) {
    await ctx.answerCbQuery();

    const tx = this.db.getTransaction(txId);

    if (!tx) {
      return ctx.reply('❌ Transaction not found.');
    }

    // Check if user is involved in transaction
    if (tx.buyer_id !== ctx.from.id && tx.seller_id !== ctx.from.id && !ctx.isAdmin) {
      return ctx.reply('⛔ You are not authorized to view this transaction.');
    }

    const role = tx.buyer_id === ctx.from.id ? 'BUYER' : 'SELLER';
    const otherParty = role === 'BUYER' ? 
      (tx.seller_username ? `@${tx.seller_username}` : tx.seller_first_name) :
      (tx.buyer_username ? `@${tx.buyer_username}` : tx.buyer_first_name);

    const statusEmoji = this.getStatusEmoji(tx.status);

    let message = `${statusEmoji} *Transaction #${tx.id}*\n\n` +
      `*Your Role:* ${role}\n` +
      `*${role === 'BUYER' ? 'Seller' : 'Buyer'}:* ${otherParty}\n\n` +
      `*Product:* ${tx.product_title}\n` +
      `*Amount:* $${tx.amount.toFixed(2)}\n` +
      `*Status:* ${tx.status.toUpperCase()}\n\n` +
      `*Timeline:*\n` +
      `Created: ${new Date(tx.created_at).toLocaleString()}\n`;

    if (tx.updated_at !== tx.created_at) {
      message += `Updated: ${new Date(tx.updated_at).toLocaleString()}\n`;
    }

    if (tx.completed_at) {
      message += `Completed: ${new Date(tx.completed_at).toLocaleString()}\n`;
    }

    message += '\n*Progress:*\n';
    message += tx.payment_confirmed ? '✅' : '⬜️';
    message += ' Payment Confirmed\n';
    message += tx.delivery_confirmed ? '✅' : '⬜️';
    message += ' Delivery Confirmed\n';
    message += tx.admin_approved ? '✅' : '⬜️';
    message += ' Admin Approved\n';

    if (tx.dispute_reason) {
      message += `\n⚠️ *Dispute Reason:*\n${tx.dispute_reason}`;
    }

    // Action buttons based on status and role
    const buttons = this.getTransactionButtons(tx, role, ctx.isAdmin);

    await ctx.reply(message, { parse_mode: 'Markdown', ...buttons });
  }

  // Get action buttons for transaction
  getTransactionButtons(tx, role, isAdmin) {
    const buttons = [];

    if (tx.status === 'pending' && role === 'BUYER' && !tx.payment_confirmed) {
      buttons.push([Markup.button.callback('✅ Confirm Payment', `confirm_payment_${tx.id}`)]);
    }

    if (tx.status === 'paid' && role === 'SELLER' && !tx.delivery_confirmed) {
      buttons.push([Markup.button.callback('📦 Confirm Delivery', `confirm_delivery_${tx.id}`)]);
    }

    if (tx.status === 'delivered' && role === 'BUYER') {
      buttons.push([Markup.button.callback('✅ Confirm Receipt', `confirm_delivery_${tx.id}`)]);
    }

    if (['paid', 'shipped', 'delivered'].includes(tx.status) && !tx.dispute_reason) {
      buttons.push([Markup.button.callback('⚠️ Open Dispute', `dispute_${tx.id}`)]);
    }

    if (isAdmin && tx.status === 'delivered' && !tx.admin_approved) {
      buttons.push([
        Markup.button.callback('✅ Approve Release', `admin_approve_${tx.id}`),
        Markup.button.callback('❌ Reject', `admin_reject_${tx.id}`)
      ]);
    }

    return buttons.length > 0 ? Markup.inlineKeyboard(buttons) : {};
  }

  // Confirm payment
  async confirmPayment(ctx, txId) {
    await ctx.answerCbQuery();

    const tx = this.db.getTransaction(txId);

    if (!tx) {
      return ctx.reply('❌ Transaction not found.');
    }

    if (tx.buyer_id !== ctx.from.id) {
      return ctx.reply('⛔ Only the buyer can confirm payment.');
    }

    if (tx.payment_confirmed) {
      return ctx.reply('✅ Payment already confirmed.');
    }

    this.db.confirmPayment(txId);

    // Notify seller
    const buyerInfo = ctx.from.username ? `@${ctx.from.username}` : ctx.from.first_name;
    this.bot.telegram.sendMessage(
      tx.seller_id,
      `💰 *Payment Confirmed!*\n\n` +
      `Transaction #${txId}\n` +
      `Product: ${tx.product_title}\n` +
      `Amount: $${tx.amount.toFixed(2)}\n\n` +
      `${buyerInfo} has confirmed payment.\n` +
      `Please ship the product and confirm delivery.`,
      { 
        parse_mode: 'Markdown',
        ...Markup.inlineKeyboard([
          [Markup.button.callback('📦 Confirm Delivery', `confirm_delivery_${txId}`)],
          [Markup.button.callback('📋 View Details', `tx_${txId}`)]
        ])
      }
    ).catch(err => console.error('Failed to notify seller:', err));

    await ctx.reply(
      '✅ *Payment Confirmed!*\n\n' +
      `Transaction #${txId}\n\n` +
      'The seller has been notified.\n' +
      'You will be notified when the product is shipped.',
      { parse_mode: 'Markdown' }
    );
  }

  // Confirm delivery
  async confirmDelivery(ctx, txId) {
    await ctx.answerCbQuery();

    const tx = this.db.getTransaction(txId);

    if (!tx) {
      return ctx.reply('❌ Transaction not found.');
    }

    // Seller confirms shipped, buyer confirms received
    if (tx.seller_id === ctx.from.id && !tx.delivery_confirmed) {
      this.db.confirmDelivery(txId);

      // Notify buyer
      const sellerInfo = ctx.from.username ? `@${ctx.from.username}` : ctx.from.first_name;
      this.bot.telegram.sendMessage(
        tx.buyer_id,
        `📦 *Product Shipped!*\n\n` +
        `Transaction #${txId}\n` +
        `Product: ${tx.product_title}\n\n` +
        `${sellerInfo} has confirmed shipment.\n` +
        `Please confirm when you receive the product.`,
        { 
          parse_mode: 'Markdown',
          ...Markup.inlineKeyboard([
            [Markup.button.callback('✅ Confirm Receipt', `confirm_delivery_${txId}`)],
            [Markup.button.callback('📋 View Details', `tx_${txId}`)]
          ])
        }
      ).catch(err => console.error('Failed to notify buyer:', err));

      // Notify admin
      const config = require('../config/config');
      for (const adminId of config.adminIds) {
        this.bot.telegram.sendMessage(
          adminId,
          `📦 *Delivery Confirmation Pending*\n\n` +
          `Transaction #${txId}\n` +
          `Product: ${tx.product_title}\n` +
          `Amount: $${tx.amount.toFixed(2)}\n\n` +
          `Seller has shipped. Waiting for buyer confirmation.`,
          { 
            parse_mode: 'Markdown',
            ...Markup.inlineKeyboard([
              [Markup.button.callback('📋 View Details', `tx_${txId}`)]
            ])
          }
        ).catch(err => console.error('Failed to notify admin:', err));
      }

      await ctx.reply(
        '✅ *Shipment Confirmed!*\n\n' +
        `Transaction #${txId}\n\n` +
        'The buyer has been notified.\n' +
        'Funds will be released after buyer confirms receipt and admin approves.',
        { parse_mode: 'Markdown' }
      );
    } else if (tx.buyer_id === ctx.from.id && tx.delivery_confirmed) {
      // Buyer confirms receipt
      this.db.updateTransactionStatus(txId, 'delivered');

      // Notify seller and admin
      const buyerInfo = ctx.from.username ? `@${ctx.from.username}` : ctx.from.first_name;
      
      this.bot.telegram.sendMessage(
        tx.seller_id,
        `✅ *Delivery Confirmed!*\n\n` +
        `Transaction #${txId}\n` +
        `Product: ${tx.product_title}\n\n` +
        `${buyerInfo} has confirmed receipt.\n` +
        `Waiting for admin to release funds.`,
        { 
          parse_mode: 'Markdown',
          ...Markup.inlineKeyboard([
            [Markup.button.callback('📋 View Details', `tx_${txId}`)]
          ])
        }
      ).catch(err => console.error('Failed to notify seller:', err));

      // Notify admin
      const config = require('../config/config');
      for (const adminId of config.adminIds) {
        this.bot.telegram.sendMessage(
          adminId,
          `✅ *Ready for Fund Release*\n\n` +
          `Transaction #${txId}\n` +
          `Product: ${tx.product_title}\n` +
          `Amount: $${tx.amount.toFixed(2)}\n\n` +
          `Both parties confirmed. Please approve fund release.`,
          { 
            parse_mode: 'Markdown',
            ...Markup.inlineKeyboard([
              [Markup.button.callback('✅ Approve', `admin_approve_${txId}`)],
              [Markup.button.callback('📋 Details', `tx_${txId}`)]
            ])
          }
        ).catch(err => console.error('Failed to notify admin:', err));
      }

      await ctx.reply(
        '✅ *Receipt Confirmed!*\n\n' +
        `Transaction #${txId}\n\n` +
        'Thank you for confirming receipt!\n' +
        'Admin will review and release funds to the seller.',
        { parse_mode: 'Markdown' }
      );
    } else {
      return ctx.reply('⚠️ Cannot perform this action at this time.');
    }
  }

  // Start dispute
  async startDispute(ctx, txId) {
    await ctx.answerCbQuery();

    const tx = this.db.getTransaction(txId);

    if (!tx) {
      return ctx.reply('❌ Transaction not found.');
    }

    if (tx.buyer_id !== ctx.from.id && tx.seller_id !== ctx.from.id) {
      return ctx.reply('⛔ You are not authorized to dispute this transaction.');
    }

    if (tx.status === 'completed' || tx.status === 'cancelled') {
      return ctx.reply('❌ This transaction is already closed.');
    }

    if (tx.dispute_reason) {
      return ctx.reply('⚠️ A dispute is already open for this transaction.');
    }

    this.db.setSessionState(ctx.from.id, 'dispute_form_reason', {
      transactionId: txId
    });

    await ctx.reply(
      '⚠️ *Open Dispute*\n\n' +
      `Transaction #${txId}\n\n` +
      'Please describe the issue in detail.\n' +
      'An admin will review your dispute and mediate.\n\n' +
      'Type your reason:',
      { 
        parse_mode: 'Markdown',
        ...Markup.keyboard([['❌ Cancel']]).resize()
      }
    );
  }
}

module.exports = TransactionHandler;

File: src/handlers/transaction_handler.py
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import ContextTypes, ConversationHandler
from src.database import Database
from src.config import Config

# Conversation states
SELLER_ID, PRODUCT_DESC, AMOUNT, DELIVERY_TIME = range(4)

async def new_transaction(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Start new transaction conversation"""
    await update.message.reply_text(
        "🛒 **Create New Transaction**\n\n"
        "Please enter the seller's Telegram ID or username:\n"
        "(You can find their ID by forwarding their message to @userinfobot)\n\n"
        "Send /cancel to abort.",
        parse_mode='Markdown'
    )
    return SELLER_ID

async def get_seller_id(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Get seller ID"""
    seller_input = update.message.text.strip()
    
    # Remove @ if present
    if seller_input.startswith('@'):
        seller_input = seller_input[1:]
    
    # Try to convert to int if it's a numeric ID
    try:
        seller_id = int(seller_input)
    except ValueError:
        await update.message.reply_text(
            "❌ Please provide a valid numeric Telegram ID.\n"
            "You can get the seller's ID by forwarding their message to @userinfobot"
        )
        return SELLER_ID
    
    # Validate seller exists and is not the buyer
    if seller_id == update.effective_user.id:
        await update.message.reply_text("❌ You cannot create a transaction with yourself!")
        return SELLER_ID
    
    context.user_data['seller_id'] = seller_id
    
    await update.message.reply_text(
        "📦 **Product Description**\n\n"
        "Please describe what you're purchasing:\n"
        "(Be detailed - include model, specifications, condition, etc.)",
        parse_mode='Markdown'
    )
    return PRODUCT_DESC

async def get_product_description(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Get product description"""
    description = update.message.text.strip()
    
    if len(description) < 10:
        await update.message.reply_text(
            "❌ Please provide a more detailed description (at least 10 characters)."
        )
        return PRODUCT_DESC
    
    context.user_data['product_description'] = description
    
    await update.message.reply_text(
        "💰 **Transaction Amount**\n\n"
        "Enter the amount (numeric value only):\n"
        "Example: 150 or 99.99",
        parse_mode='Markdown'
    )
    return AMOUNT

async def get_amount(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Get transaction amount"""
    try:
        amount = float(update.message.text.strip())
        if amount <= 0:
            raise ValueError("Amount must be positive")
    except ValueError:
        await update.message.reply_text(
            "❌ Please enter a valid positive number.\n"
            "Example: 150 or 99.99"
        )
        return AMOUNT
    
    context.user_data['amount'] = amount
    
    await update.message.reply_text(
        "⏱️ **Expected Delivery Time**\n\n"
        "When should the product be delivered?\n"
        "Example: 3 days, 1 week, 2-3 business days",
        parse_mode='Markdown'
    )
    return DELIVERY_TIME

async def get_delivery_time(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Get delivery time and create transaction"""
    delivery_time = update.message.text.strip()
    
    if len(delivery_time) < 3:
        await update.message.reply_text("❌ Please provide a valid delivery timeframe.")
        return DELIVERY_TIME
    
    # Create transaction
    db = Database()
    transaction_id = await db.create_transaction(
        buyer_id=update.effective_user.id,
        seller_id=context.user_data['seller_id'],
        product_description=context.user_data['product_description'],
        amount=context.user_data['amount'],
        delivery_time=delivery_time
    )
    
    # Prepare summary
    summary = f"""
✅ **Transaction Created Successfully!**

**Transaction ID:** #{transaction_id}
**Seller ID:** {context.user_data['seller_id']}
**Product:** {context.user_data['product_description']}
**Amount:** ${context.user_data['amount']:.2f}
**Expected Delivery:** {delivery_time}
**Status:** 🟡 PENDING (Awaiting admin approval)

An admin will review your transaction shortly.
You'll be notified once it's approved! 🔔
"""
    
    await update.message.reply_text(summary, parse_mode='Markdown')
    
    # Notify admin
    try:
        admin_message = f"""
🔔 **New Transaction Created**

**ID:** #{transaction_id}
**Buyer:** {update.effective_user.first_name} (ID: {update.effective_user.id})
**Seller ID:** {context.user_data['seller_id']}
**Product:** {context.user_data['product_description']}
**Amount:** ${context.user_data['amount']:.2f}
**Delivery:** {delivery_time}

Use /admin to review and approve.
"""
        await context.bot.send_message(
            chat_id=Config.ADMIN_ID,
            text=admin_message,
            parse_mode='Markdown'
        )
    except Exception as e:
        print(f"Failed to notify admin: {e}")
    
    # Clear user data
    context.user_data.clear()
    return ConversationHandler.END

async def cancel_transaction(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Cancel transaction creation"""
    context.user_data.clear()
    await update.message.reply_text(
        "❌ Transaction creation cancelled.",
        parse_mode='Markdown'
    )
    return ConversationHandler.END

async def my_transactions(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Show user's transactions as buyer"""
    db = Database()
    transactions = await db.get_user_transactions(update.effective_user.id, as_buyer=True)
    
    if not transactions:
        await update.message.reply_text(
            "📭 You don't have any active transactions.\n\n"
            "Use /newtransaction to create one!"
        )
        return
    
    await update.message.reply_text(
        f"🛒 **Your Active Transactions** ({len(transactions)})\n\n"
        "Select a transaction to view details:",
        parse_mode='Markdown'
    )
    
    for tx in transactions:
        status_emoji = {
            'PENDING': '🟡',
            'APPROVED': '🟢',
            'SHIPPED': '📦',
            'DISPUTED': '⚠️'
        }.get(tx['status'], '⚪')
        
        keyboard = [[
            InlineKeyboardButton(
                f"{status_emoji} Transaction #{tx['id']} - ${tx['amount']:.2f}",
                callback_data=f"view_tx_{tx['id']}"
            )
        ]]
        reply_markup = InlineKeyboardMarkup(keyboard)
        
        await update.message.reply_text(
            f"**Product:** {tx['product_description'][:50]}...",
            reply_markup=reply_markup,
            parse_mode='Markdown'
        )

async def my_sales(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Show user's transactions as seller"""
    db = Database()
    transactions = await db.get_user_transactions(update.effective_user.id, as_buyer=False)
    
    if not transactions:
        await update.message.reply_text(
            "📭 You don't have any active sales.\n\n"
            "Waiting for buyers to create transactions with you!"
        )
        return
    
    await update.message.reply_text(
        f"💼 **Your Active Sales** ({len(transactions)})\n\n"
        "Select a sale to view details:",
        parse_mode='Markdown'
    )
    
    for tx in transactions:
        status_emoji = {
            'PENDING': '🟡',
            'APPROVED': '🟢',
            'SHIPPED': '📦',
            'DISPUTED': '⚠️'
        }.get(tx['status'], '⚪')
        
        keyboard = [[
            InlineKeyboardButton(
                f"{status_emoji} Sale #{tx['id']} - ${tx['amount']:.2f}",
                callback_data=f"view_sale_{tx['id']}"
            )
        ]]
        reply_markup = InlineKeyboardMarkup(keyboard)
        
        await update.message.reply_text(
            f"**Product:** {tx['product_description'][:50]}...",
            reply_markup=reply_markup,
            parse_mode='Markdown'
        )
